src/neuralnet/superblock.cpp

```diff
@@ -177,38 +208,299 @@ class SuperblockValidator
+    //!
+    //! Nodes validate superblocks in by-project fallback convergence cases by
+    //! matching manifest project parts to the convengence hints embedded in a
+    //! superblock project section. These hints are truncated SHA256 hashes so
+    //! a hint can qualify more than one manifest part for consideration while
+    //! reconstructing a convergence for validation.
+    //!
+    //! This class behaves like a cursor through each combination of the parts
+    //! it is initialized with. For each call to \c GetNextConvergence(), this
+    //! class generates a convergence using a new combination of project parts
+    //! until it exhausts every combination or until the client code matches a
+    //! superblock with one of the results.
+    //!
+    //! We have built here a catapult that fires a huge boulder to kill a tiny
+    //! bird--the chance of a project part hash collision resolved by the hint
+    //! in the superblock is incredibly small, but this provision prevents the
+    //! validation of a superblock from failing if a collision ever occurs.
+    //!
+    class ProjectCombiner
+    {
+    public:
+        //!
+        //! \brief Initialize a project combiner with the provided collection
+        //! of resolved projects.
+        //!
+        //! \param projects Contains matching manifest parts hashes for each
+        //! project hinted in the superblock.
+        //!
+        ProjectCombiner(std::map<std::string, ResolvedProject> projects)
+            : m_projects(std::move(projects))
+            , m_total_combinations(1)
+            , m_current_combination(0)
+        {
+            for (auto& project_pair : reverse_iterate(m_projects)) {
+                project_pair.second.m_combiner_mask = m_total_combinations;
+                m_total_combinations *= project_pair.second.m_resolved_parts.size();
+            }
+        }
+
+        //!
+        //! \brief Initialze a project combiner that produces no results.
+        //!
+        ProjectCombiner()
+            : m_total_combinations(0)
+            , m_current_combination(0)
+        {
+        }
+
+        //!
+        //! \brief Get the total number of possible project part combinations
+        //! that the object can create convergences from.
+        //!
+        //! \return
+        //!
+        size_t TotalCombinations() const
+        {
+            return m_total_combinations;
+        }
+
+        //!
+        //! \brief Generate the next convergence in sequence from the resolved
+        //! project parts.
+        //!
+        //! Jim Owens enlightens us with a stimulating description of his algorithm:
+        //!
+        //! Compute number of permutations and cumulative placevalue. For instance, if there were 4 projects, and
+        //! the number of candidate parts for each of the projects were 2, 3, 2, and 2, then the total number of
+        //! permutations would be 2 * 3 * 2 * 1 = 12, and the cumulative placevalue would be
```

Comment:
I did a typo here in my original comments. I changed it midstream to be 2, 3, 2, 2 and forgot to update that part... it should be 2 * 3 * 2 * 2 = 24

Reply:
Fixed, along with a couple of other minor doc comments and tweaks 🙂

Reply:
Merging!